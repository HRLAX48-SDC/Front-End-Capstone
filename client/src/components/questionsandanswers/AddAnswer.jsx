import React, {useState} from 'react';
import axios from 'axios';
import { Button, Modal, closeButton, Form, Row, Col, FloatingLabel, Image} from 'react-bootstrap';
import API_KEY from '../../config/config.js';

const AddAnswer = (props) => {
  const [showModal, setModal] = useState(false);
  const handleShow = () => setModal(true);
  const handleClose = () => setModal(false);
  const [body, setBody] = useState('');
  const [name, setName] = useState('');
  const [email, setEmail] = useState('');
  const [photos, setphotos] = useState([]);
  const [file, setFile] = useState('');

  const onFileChange = (e) => {
    const preview = document.querySelector('#q_ans_uploads_previews');
    const file = e.target.files[0];
    setFile(file);
    const reader = new FileReader();
    reader.onload = () => {
      preview.src = reader.result;
    };
    if(file) {
      reader.readAsDataURL(file);
    };
  };
  const uploadImage = (e) => {
    e.preventDefault();
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', 'rp7rer9y');
    fetch('https://api.cloudinary.com/v1_1/dtnikmimx/image/upload', {
      method: 'POST',
      body: formData,
    })
    .then(res => res.json())
    .then((res) => {
      //console.log(res.url);
      alert('success upload')
      setphotos([res.url]);
    });
  }

  const postAnswer = (e) => {
    //e.preventDefault();
    const answer_info = {
      body: body,
      name: name,
      email: email,
      photos: photos
    }
    if(body.length > 1 && name.length > 1 && email.length > 1 && email.includes('@')) {
      axios.post(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/qa/questions/${props.question_id}/answers`, answer_info, {
        headers: { 'Authorization': `${API_KEY}` }
      })
        .then(response => {
          //console.log(response);
          alert('your question is successfully post.');
          props.getAllAnswers();
          handleClose();
        })
        .catch((err) => console.error(err));
    } else {
      if(!email.includes('@')) {
        alert('Please check your email format.')
      } else {
        alert('You must enter the all mandatory field.');
      };
    };
  };

  return (
    <div id='add_answer_button'>
      {' '}
      <Button className='answer_add' variant="outline-secondary" size='sm' onClick={handleShow}>ADD A ANSWER</Button>
      <Modal show={showModal} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>
            <h3 id='question_modal_title'>Submit your Answer</h3>
            <h6 id='question_modal_subtitle'>{props.product_name}: {props.question_body}</h6>
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form>
            <Row className='mb-3'>
              <Form.Group as={Col} controlId='formGridQuestion'>
                <Form.Label>Post your answer here</Form.Label>
                <Form.Control
                  as='textarea'
                  type='text'
                  name='body'
                  placeholder='Please enter a answer.'
                  style={{ height: '100px'}}
                  maxLength='1000'
                  onChange={(e) => setBody(e.target.value)}
                  className='add_answer_body_input'
                  value={body}
                  ></Form.Control>
              </Form.Group>
            </Row>
            <Row className='mb-3'>
              <Form.Group as={Col} controlId='formGridNickname'>
                <Form.Label>What is your nickname?</Form.Label>
                <FloatingLabel controlId='floatingnickname' label='Example: jackson543!'>
                  <Form.Control
                    type='text'
                    name='name'
                    placeholder='Example: jackson543!'
                    maxLength='60'
                    onChange={(e) => setName(e.target.value)}
                    className='add_answer_name_input'
                    value={name}
                  ></Form.Control>
                </FloatingLabel>
                <Form.Text className='text-muted'>
                  For privacy reasons, do not use your full name or email address
                </Form.Text>
              </Form.Group>
            </Row>
            <Row className='mb-3'>
              <Form.Group as={Col} controlId='formGridNickname'>
                <Form.Label>What is your email?</Form.Label>
                <FloatingLabel controlId='floatingemail' label='Example: jack@email.com'>
                  <Form.Control
                    type='email'
                    name='email'
                    placeholder='Example: jack@email.com'
                    maxLength='60'
                    onChange={(e) => setEmail(e.target.value)}
                    className='add_answer_email_input'
                    value={email}
                  ></Form.Control>
                </FloatingLabel>
                <Form.Text className='text-muted'>
                  For authentication reasons, you will not be emailed
                </Form.Text>
              </Form.Group>
            </Row>
            <Row>
              <Form.Group as={Col} controlId='formGridphotos'>
                <Form.Label>Upload photos here</Form.Label>
                <Form.Control
                  type='file'
                  accept='image/png, image/jpeg'
                  onChange={onFileChange}/>
                <br/>
                <Image id='q_ans_uploads_previews' height='100' width='100'/>
              </Form.Group>
            </Row>
            <Row>
              <Form.Group>
                <Button className='addA_upload_btn' type='submit' onClick={uploadImage} size='sm' variant="light">Upload Image</Button>
              </Form.Group>
            </Row>
          </Form>
        </Modal.Body>
        <Modal.Footer>
          <Button className='addA_cancel_btn' variant='secondary' onClick={() => handleClose()}>Cancel</Button>
          <Button className='addA_submit_btn' type='submit' onClick={postAnswer}>Submit</Button>
        </Modal.Footer>
      </Modal>
    </div>
  );
};
export default AddAnswer;