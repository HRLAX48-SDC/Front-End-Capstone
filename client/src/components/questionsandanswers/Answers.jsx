import React, {useState, useEffect} from 'react';
import axios from 'axios';
import API_KEY from '../../config/config.js';
import IndividualAnswer from './IndividualAnswer.jsx';
import { Container, Row, Col, Stack, Button } from 'react-bootstrap';
import AddAnswer from './AddAnswer.jsx';

const Answers = (props) => {
  const [question_id, setQId] = useState(() => props.question.question_id);
  useEffect(() => {
    if(props.question.question_id) {
      setQId(props.question.question_id);
    };
  }, [props.question.question_id]);
  const [answers, setAnswers] = useState([]);
  useEffect(() => {
    getAllAnswers();
  }, [question_id]);
  const getAllAnswers = () => {
    axios.get(`https://app-hrsei-api.herokuapp.com/api/fec2/hr-lax/qa/questions/${question_id}/answers?page=1&count=500`, {
    headers: { 'Authorization': `${API_KEY}` }
    })
      .then((response) => {
        setAnswers(response.data.results);
      })
      .catch((err) => console.error(err));
  }
  const ansLength = answers.length;
  const [ansPerPage, setAnsPerPage] = useState(2);
  const loadMoreAns = () => {
    setAnsPerPage(ansPerPage + 2);
  }
  const collapse = () => {
    setAnsPerPage(2);
  }
  const sliceAns = answers.slice(0, ansPerPage);
  const changeBtnMode = () => {
    if(ansPerPage < ansLength) {
      return(
        <strong onClick={() => loadMoreAns()}>LOAD MORE ANSWERS({answers.length-sliceAns.length})</strong>
      );
    } else {
      return(
        <strong id='collapseBtn' variant="outline-secondary" onClick={()=> collapse()}>COLLAPSE ANSWERS</strong>
      );
    };
  };
  return (
    <div>
      {answers.length > 0 &&
      <Row id='q_grid_row2'>
        <Col id='q_grid_row2_col1' xs={1}><strong>A:</strong></Col>
        <Col className="q_answerslist" id='q_grid_row2_col2'>
          {sliceAns.map((answer, index) =>
            <IndividualAnswer answer={answer} key={answer.answer_id} />
          )}
          <Col>
            <Stack direction='horizontal' gap={2}>
              {answers.length > 2 && changeBtnMode()}
              <AddAnswer question_id={question_id} question_body={props.question.question_body} product_name={props.product_name} getAllAnswers={getAllAnswers}/>
            </Stack>
          </Col>
        </Col>
      </Row>
      }
      {/* <br></br> */}
    </div>
  )

};
export default Answers;