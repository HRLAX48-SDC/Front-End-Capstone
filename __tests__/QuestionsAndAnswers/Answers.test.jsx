import React from 'react';
import Enzyme, {shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Answers from '../../client/src/components/questionsandanswers/Answers.jsx';
import AddAnswer from '../../client/src/components/questionsandanswers/AddAnswer.jsx';
Enzyme.configure({adapter: new Adapter()});

const zeroQuestion = [];
const oneQuestion = [{
  question_id: 426448,
  question_body: 'Is it water proof?',
  question_date: "2021-09-24T00:00:00.000Z",
  question_helpfulness: 6,
  asker_name: "username123",
  answers: []
  }];

describe('<Answers /> with props', () => {
  test('Render an empty answer component', () => {
    const wrapper = shallow(<Answers question={zeroQuestion}/>);
    expect(wrapper.find('IndividualAnswer')).toHaveLength(0);
  });
  test('Render answer compoment correctly', () => {
    const wrapper = renderer.create(<Answers question={oneQuestion}/>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Test add answer button', () => {
  test('Render an add Answer button', () => {
    const wrapper = mount(<Answers question={oneQuestion}/>);
    expect(wrapper).toContainMatchingElement('AddAnswer');
  })
})