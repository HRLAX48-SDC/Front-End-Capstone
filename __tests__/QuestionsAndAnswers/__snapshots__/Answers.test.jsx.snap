// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Answers /> with props Render answer compoment correctly 1`] = `
<div>
  <div
    className="row"
    id="q_grid_row2"
  >
    <div
      className="col-1"
      id="q_grid_row2_col1"
    >
      <strong>
        A:
      </strong>
    </div>
    <div
      className="q_answerslist col"
      id="q_grid_row2_col2"
    >
      <div
        className="col"
      >
        <div
          className="hstack gap-2"
        >
          <div
            id="add_answer_button"
          >
             
            <button
              className="answer_add btn btn-outline-secondary btn-sm"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              ADD A ANSWER
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
